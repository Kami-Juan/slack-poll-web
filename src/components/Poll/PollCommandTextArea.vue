<template>
  <div id="poll-command-text-area">
    <b-row align-h="center">
      <b-col cols="12" md="6">
        <span id="instructions">Copy-pasta your command into slack :^)</span>
        <textarea
          :value="value"
          @input="$emit('input', $event.target.value)"
          id="command-input"
        >
        </textarea>
      </b-col>
    </b-row>
  </div>
</template>

<script>
export default {
  name: 'PollCommandTextArea',
  props: ['value']
}
</script>

<style>
#command-input {
  width: 100%;
  height: 92px;
  outline: 0;
  border: 1px solid var(--secondary-border-color);
  border-radius: 4px;
  resize: none;
  padding: 10px 15px 15px 15px;
}

#poll-command-text-area {
  text-align: left;
}
</style>
