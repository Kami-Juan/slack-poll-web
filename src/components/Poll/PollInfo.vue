<template>
  <div id="poll-info">
    <div id="bot-avatar"></div>
    <div id="bot-message">
      <h4 id="bot-name">Yellow Poll</h4>
      <slack-section>
        <div id="info-group">
          <h4 id="poll-title">{{ title }}</h4>
          <div>By: You</div>
          <div>
            Mode:
            <a id="poll-type" @click="togglePollType">{{ multiple ? 'Multiple' : 'Single' }}</a>
          </div>
        </div>
      </slack-section>
      <div id="options-slot">
        <slot></slot>
      </div>
    </div>
  </div>
</template>

<script>
import SlackSection from '../Decoration/SlackSection'

export default {
  name: 'PollInfo',
  components: {
    SlackSection
  },
  props: ['title'],
  methods: {
    togglePollType() {
      this.multiple = !this.multiple
      this.$emit('typeChanged', this.multiple)
      this.$emit('updated')
    }
  },
  data() {
    return {
      multiple: false
    }
  }
}
</script>

<style>
#poll-info {
  margin-bottom: 10px;
  align-items: center;
  display: flex;
  flex-direction: row;
  align-items: start;
}

#poll-title {
  font-family: LatoBold;
  font-size: 19px;
  font-weight: bold;
  margin-top: 0;
  margin-bottom: 5px;
}

#poll-type {
  font-family: LatoBold;
  font-weight: bold;
  text-decoration: underline;
}

#bot-avatar {
  width: 40px;
  height: 40px;
  border-radius: 3px;
  background-color: var(--yellowme-color);
  margin-right: 10px;
}

#bot-name {
  font-family: LatoBold;
  font-size: 19px;
  font-weight: bold;
  margin-top: 0;
  margin-bottom: 2px;
  text-align: left;
}

#options-slot {
  margin-top: 10px;
}
</style>