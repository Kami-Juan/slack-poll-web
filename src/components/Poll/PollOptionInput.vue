<template>
  <b-row align-h="center" id="poll-option-input">
    <b-col cols=6>
      <input
        type="text"
        :value="value"
        @input="$emit('input', $event.target.value)"
        @keyup.enter="$emit('keyupEnter')"
        id="add-option-input"
      >
    </b-col>
  </b-row>
</template>

<script>
export default {
  name: 'PollOptioninput',
  props: ['value']
}
</script>

<style>
#add-option-input {
  width: 100%;
}
</style>
