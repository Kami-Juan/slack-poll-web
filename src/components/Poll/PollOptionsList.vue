<template>
  <div id="poll-options-list">
    <b-row align-h="center">
      <b-col cols=6>
        <ul id="options-list">
          <li v-for="option in options" :key="option.id">
            <input type="text" :value="option.value">
            <a @click="$emit('click')">‚ùå</a>
          </li>
        </ul>
      </b-col>
    </b-row>
  </div>
</template>

<script>
export default {
  name: 'PollOptionsList',
  props: {
    options: {
      type: Object,
      required: true
    }
  }
}
</script>

<style>
</style>
