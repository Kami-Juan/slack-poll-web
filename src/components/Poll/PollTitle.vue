<template>
  <div id="poll-title">
    <b-row align-h="center">
      <b-col cols=6>
        <input
          id="title-input"
          type="text"
          placeholder="Here goes your poll title"
          autocomplete="off"
          :value="value"
          @input="$emit('input', $event.target.value)"
          @keyup.enter="$emit('keyupEnter')"
        >
      </b-col>
    </b-row>
  </div>
</template>

<script>
export default {
  name: 'PollTitle',
  props: ['value']
}
</script>

<style>
#title-input {
  width: 100%;
  outline: none;
  border: 0;
  font-size: 48px;
}

#title-input {
  outline: none;
}
</style>
