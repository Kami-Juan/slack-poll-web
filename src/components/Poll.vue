<template>
  <div class="Poll">
    <b-container>
      <poll-title v-model="title" @keyupEnter="focusOnNextInput" />
      <poll-option-input v-model="newOptionText" @keyupEnter="addOption" />
      <b-row align-h="center" id="slack-message-container">
        <b-col cols="6">
          <poll-info :title="title" @typeChanged="typeChanged($event)">
            <slack-section>
              <ul id="options-list">
                <li v-for="option in options" :key="option.id" class="option-lsit-item">
                  <!-- <input type="text" :value="option.value" class="option-input-element"> -->
                  <span class="option-element">{{ option.value }}</span>
                  <a class="option-remove-element" @click="removeOption(option)">❌</a>
                </li>
              </ul>
            </slack-section>
          </poll-info>
        </b-col>
      </b-row>
      <poll-command-text-area v-model="command" />
      <b-row align-h="center">
        <b-col cols="3">
          <a
            :class="successfulCopy ? 'text-success' : ''"
            @click="copyToClipboard"
          >
            {{ successfulCopy ? 'Copied!' : 'Copy' }}
          </a>
        </b-col>
      </b-row>
      <!-- <b-modal id="success-modal" title="Se copió el comando con éxito">
        <p class="my-4">¡Ahora ve a pegarlo en slack!</p>
      </b-modal> -->
    </b-container>
  </div>
</template>

<script>
import { transcript } from './utils/transcriptor'
import PollTitle from './Poll/PollTitle'
import PollInfo from './Poll/PollInfo'
import PollOptionInput from './Poll/PollOptionInput'
import PollCommandTextArea from './Poll/PollCommandTextArea'
import SlackSection from  './Decoration/SlackSection'

export default {
  name: 'Poll',
  components: {
    PollTitle,
    PollInfo,
    PollOptionInput,
    PollCommandTextArea,
    SlackSection
  },
  methods: {
    addOption() {
      let id = 1
      if (this.options.length > 0) {
        id = this.options[0].id++
      }
      this.options.push({
        id: id,
        value: this.newOptionText
      })
      this.newOptionText = ''
    },

    focusOnNextInput() {
      document.getElementById('add-option-input').focus()
    },

    removeOption(option) {
      this.options.splice(this.options.indexOf(option), 1)
    },

    copyToClipboard() {
      let command = document.querySelector('#command-input')
      command.select()
      try {
        document.execCommand('copy')
        this.successfulCopy = true
      } catch {
        alert('Oops!')
      }
    },

    typeChanged(event) {
      this.multiple = event
      this.updateCommand()
    },

    updateCommand() {
      this.command = transcript({
        title: this.title,
        options: this.options.map(option => option.value),
        multiple: this.multiple
      })
    }
  },
  watch: {
    command() {
      this.successfulCopy = false
    }
  },
  data() {
    return {
      title: '',
      newOptionText: '',
      options: [],
      multiple: false,
      command: '',
      successfulCopy: false
    }
  },
  updated() {
    this.updateCommand()
  }
}
</script>

<style scoped>
#title-input,
#add-option-input,
#command-input {
  width: 100%;
}

#options-list {
  text-align: left;
  list-style-type: none;
  margin: 0;
  padding: 0;
}

#slack-message-container {
  margin-top: 40px;
  margin-bottom: 40px;
}

/*This is for a input type*/
/* 
.option-input-element {
  border: none;
}

.option-input-element:focus {
  outline: 0;
}
*/

.option-element {
  margin-right: 12px;
}

.option-remove-element {
  cursor: pointer;
}
</style>
