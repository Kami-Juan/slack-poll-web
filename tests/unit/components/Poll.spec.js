import BootstrapVue from 'bootstrap-vue'
import { LayoutPlugin, ModalPlugin } from 'bootstrap-vue'

import { shallowMount, createLocalVue, mount } from '@vue/test-utils'

import Poll from './../../../src/components/Poll'
import PollOptionInput from './../../../src/components/Poll/PollOptionInput'
import PollCommandTextArea from './../../../src/components/Poll/PollCommandTextArea'


describe('test Poll.vue', () => {
  const localVue = createLocalVue()

  localVue.use(BootstrapVue)
  localVue.use(LayoutPlugin)
  localVue.use(ModalPlugin)

  beforeEach(() => {
    localStorage.clear()
  })

  it('should be render', () => {
    const cmp = mount(Poll, {localVue, components: { PollOptionInput }})
    expect(cmp.element).toMatchSnapshot()
  })

  it('set default data if in the localstorage don"t have information', async () => {
    const cmp = shallowMount(Poll, {localVue})

    await cmp.vm.$nextTick()

    expect(cmp.vm.title).toBe('')
    expect(cmp.vm.options).toStrictEqual([])
    expect(cmp.vm.multiple).toBe(false)
  })

  it('should be setted data from the localstorage', async () => {
    localStorage.setItem('title', 'test')
    localStorage.setItem('options', '[{"value": "frijoles"}]')
    localStorage.setItem('multiple', 'true')

    const cmp = shallowMount(Poll, {localVue})

    await cmp.vm.$nextTick()

    expect(cmp.vm.title).toBe('test')
    expect(cmp.vm.options).toStrictEqual([{ value: 'frijoles' }])
    expect(cmp.vm.multiple).toBe(true)
  })

  it('doesn"t add a option if doesn"t have information to added', async () => {
    const cmp = shallowMount(Poll, {localVue})

    cmp.vm.addOption()

    expect(cmp.vm.options.length).toBe(0)
  })

  it('add a option if have information to added', async () => {
    const cmp = shallowMount(Poll, {localVue})

    cmp.setData({ newOptionText: 'Tamales' })
    cmp.vm.addOption()

    const optionsOffline = localStorage.getItem('options')

    expect(cmp.vm.options.length).toBe(1)
    expect(cmp.vm.newOptionText).toBe('')
    expect(optionsOffline).not.toBeNull()
  })

  it('remove a option and update the localstorage', async () => {
    const cmp = shallowMount(Poll, {localVue})

    cmp.setData({ options: [{ value: 'Tamales' }] })
    cmp.vm.removeOption(0)

    const optionsOffline = localStorage.getItem('options')

    expect(optionsOffline).toBe('[]')
  })

  it('update a option and change the value in the localstorage', () => {
    const cmp = shallowMount(Poll, {localVue})
    cmp.setData({ options: [{ value: 'Tamales' }] })
    cmp.vm.updateOptionPersisted('Hojaldras', 0)

    const optionsOffline = JSON.parse(localStorage.getItem('options'))

    expect(cmp.vm.options[0].value).toBe('Hojaldras')
    expect(optionsOffline.length).toBe(1)
    expect(optionsOffline[0].value).toBe('Hojaldras')
  })

  it('update a option who doesn"t exists in the list', () => {
    const cmp = shallowMount(Poll, {localVue})
    cmp.setData({ options: [{ value: 'Tamales' }] })
    cmp.vm.updateOptionPersisted('Hojaldras', 1)

    const optionsOffline = JSON.parse(localStorage.getItem('options'))

    expect(cmp.vm.options[0].value).toBe('Tamales')
    expect(optionsOffline.length).toBe(1)
    expect(optionsOffline[0].value).toBe('Tamales')
  })

  it('change the current focus to another option', async () => {
    const cmp = mount(Poll, {localVue, components: { PollOptionInput }, attachToDocument: true})

    cmp.vm.focusOnNextInput()
    const optionInput = cmp.find('#add-option-input').element

    expect(optionInput).toBe(document.activeElement)
  })

  it('no copy the command from the command-textarea because the browser no supported', () => {
    window.alert = jest.fn()
    const cmp = mount(Poll, {localVue, components: { PollCommandTextArea }, attachToDocument: true})

    cmp.vm.copyToClipboard()

    expect(window.alert).toHaveBeenCalled()
    expect(cmp.vm.successfulCopy).toBe(false)
  })

  it('copy the command from the command-textarea', () => {
    document.execCommand = jest.fn()
    const cmp = mount(Poll, {localVue, components: { PollCommandTextArea }, attachToDocument: true})

    cmp.vm.copyToClipboard()
    expect(document.execCommand).toHaveBeenCalledWith('copy')
    expect(cmp.vm.successfulCopy).toBe(true)
  })

  it('update the title when is changed', () => {
    localStorage.setItem('title', 'test 1')

    const cmp = mount(Poll, {localVue})
    cmp.setData({ title: 'test 2' })

    const titleOffline = localStorage.getItem('title')

    expect(titleOffline).toBe('test 2')
  })

  it('update the multiple value when is changed', () => {
    localStorage.setItem('multiple', 'true')

    let cmp = mount(Poll, {localVue})

    cmp.vm.updateCommand = jest.fn()
    cmp.setData({ multiple: false })

    const multipleOffline = localStorage.getItem('multiple')

    expect(multipleOffline).toBe('false')
    expect(cmp.vm.updateCommand).toBeCalled()
  })
})