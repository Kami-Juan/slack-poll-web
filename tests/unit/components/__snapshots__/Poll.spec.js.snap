// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Poll.vue should be render 1`] = `
<div
  class="Poll"
>
  <div
    class="container"
  >
    <div
      id="poll-title"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-md-6 col-12"
        >
          <input
            autocomplete="off"
            id="title-input"
            placeholder="Here goes your poll title"
            type="text"
          />
        </div>
      </div>
    </div>
     
    <div
      class="row justify-content-center"
      id="poll-option-input"
    >
      <div
        class="col-md-6 col-12"
      >
        <input
          id="add-option-input"
          placeholder="Add an amazing option"
          type="text"
        />
      </div>
    </div>
     
    <div
      class="row justify-content-center"
      id="slack-message-container"
    >
      <div
        class="col-6"
      >
        <div
          id="poll-info"
        >
          <div
            id="bot-avatar"
          />
           
          <div
            id="bot-message"
          >
            <h4
              id="bot-name"
            >
              Yellow Poll
            </h4>
             
            <div
              class="slack-section"
            >
              <div
                class="separator"
              />
               
              <div
                class="slack-slot"
              >
                <div
                  id="info-group"
                >
                  <h4
                    id="poll-title"
                  >
                    
                  </h4>
                   
                  <div
                    id="poll-username"
                  >
                    By: You
                  </div>
                   
                  <div>
                    <span
                      id="poll-mode"
                    >
                      Mode:
                    </span>
                     
                    <a
                      id="poll-type"
                    >
                      
            Single
          
                    </a>
                  </div>
                </div>
              </div>
            </div>
             
            <div
              id="options-slot"
            >
              <div
                class="slack-section"
              >
                <div
                  class="separator"
                />
                 
                <div
                  class="slack-slot"
                >
                  <ul
                    id="options-list"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      id="poll-command-text-area"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-md-6 col-12"
        >
          <span
            id="instructions"
          >
            Copy-pasta your command into slack :^)
          </span>
           
          <textarea
            id="command-input"
          />
        </div>
      </div>
    </div>
     
    <div
      class="row justify-content-center"
    >
      <div
        class="col-3"
      >
        <a
          class=""
        >
          Copy
        </a>
      </div>
    </div>
  </div>
</div>
`;
